### PR TITLE
[Web-22] add function addyear dropdown

### PR DESCRIPTION
**Pull request by emilyPhee**
https://user-images.githubusercontent.com/57051508/212405158-765e911e-8bd2-4108-ae28-bac77a5e8039.mov

- Disable option when year added
- Display the side Toast for action completion

_AppUser's years array only contains unique years value and is sorted in ascending order._

### CODE REVIEW

**Review by emilyPhee — COMMENTED** [review]
reviewed

**Comment by emilyPhee** [comment]
[[Web] Add functionalities to AddYearDropdown](https://trello.com/c/GLIQGbF8/22-web-add-functionalities-to-addyeardropdown)

File: new-nextjs-client/src/components/index-page/AddYearDropdown/AddYearDropdown.tsx [hunk 1]
```diff
@@ -24,11 +26,33 @@ const getYearOptions = (): YearOption[] => {
...
   const onSelectChange = (newValue: YearOption) => {
-    updateAppUser((draft) => {
-      draft.addYear(newValue.value);
-    });
+    if (!appUser.years.includes(newValue.value)) {
+      updateAppUser((draft) => {
+        draft.addYear(newValue.value);
+        draft.sortYear();
+
+        showToastBox({
```
Comment (anthopark) [inline new-nextjs-client/src/components/index-page/AddYearDropdown/AddYearDropdown.tsx]: call `showToastBox()` outside the `updateAppUser()`.

updateAppUser should only include code related to updating the state of AppUser

File: new-nextjs-client/src/components/index-page/AddYearDropdown/AddYearDropdown.tsx [hunk 2]
```diff
@@ -24,11 +26,33 @@ const getYearOptions = (): YearOption[] => {
...
+  const showToastBox = useAppToast();
 
   const onSelectChange = (newValue: YearOption) => {
-    updateAppUser((draft) => {
-      draft.addYear(newValue.value);
-    });
+    if (!appUser.years.includes(newValue.value)) {
+      updateAppUser((draft) => {
+        draft.addYear(newValue.value);
+        draft.sortYear();
```
Comment (anthopark) [inline new-nextjs-client/src/components/index-page/AddYearDropdown/AddYearDropdown.tsx]: sortYear() shouldn't be called here. the sorting should be handled inside `addYear()` method.

File: new-nextjs-client/src/components/index-page/AddYearDropdown/AddYearDropdown.tsx [hunk 3]
```diff
@@ -24,11 +26,33 @@ const getYearOptions = (): YearOption[] => {
...
+
+        showToastBox({
+          status: 'success',
+          highlightedData: newValue.label,
+          message: 'year added',
+          duration: 3500,
+        });
+      });
+
+      const newYearOptionsArr = yearOptions.map((obj) => {
```
Comment (anthopark) [inline new-nextjs-client/src/components/index-page/AddYearDropdown/AddYearDropdown.tsx]: Let's make a separate helper function for this. you can name `disableSelectedOptions()` then call `setYearOptions()` inside it with the updated options.

File: new-nextjs-client/src/entities/app-user.ts [hunk 4]
```diff
@@ -3,6 +3,7 @@ import { immerable } from 'immer';
 interface IAppUser {
   years: number[];
   addYear: (year: number) => void;
+  sortYear: () => void;
```
Comment (anthopark) [inline new-nextjs-client/src/entities/app-user.ts]: Remove this and make it a private method. which will be only called in `addYear()`

**Review by anthopark — CHANGES_REQUESTED** [review]
Good work. Left a few comments to be addressed

**Review by anthopark — APPROVED** [review]

**Review by anthopark — APPROVED** [review]

**Pull request merged**